### PR TITLE
[BACKLOG-3790] Add support for ranged assignment of conflicting karaf…

### PR DESCRIPTION
**Pull request by tkafalas**
… ports.  Add pentaho/api/karaf/properties web service.

### CODE REVIEW

**Comment by tkafalas** [comment]
Nick, can you take a look at this?  Thanks.

File: assembly/package-res/biserver/pentaho-solutions/system/pentahoServices.spring.xml [hunk 1]
```diff
@@ -148,4 +148,5 @@
   </bean>
 
   <bean class="org.pentaho.platform.web.http.api.resources.PasswordResource" scope="request"/>
+  <bean class="org.pentaho.platform.web.http.api.resources.KarafResource" scope="request"/>
```
Comment (pentaho-nbaker) [inline assembly/package-res/biserver/pentaho-solutions/system/pentahoServices.spring.xml]: Shouldn't be Karaf Specific I wouldn't think. I envision the endpoint returning service-> port assignments for the entire server. The fact that we're using Karaf shouldn't be apparent from the service layer

File: extensions/src/org/pentaho/platform/osgi/KarafInstancePropertyDto.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * 
+ * @author tkafalas
+ *
+ */
+@XmlRootElement(name="KarafInstanceProperty")
+@XmlAccessorType(XmlAccessType.FIELD)
+public class KarafInstancePropertyDto implements Serializable {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafInstancePropertyDto.java]: Why a separate DTO class, couldn't you just annotate the KarafInstanceProperty class?

File: extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+package org.pentaho.platform.osgi;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+
+import org.apache.commons.lang.StringUtils;
+
+public class KarafPropertyImporter {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java]: Importer already has meaning in the platform. I would rename is KarafPropertyParser or KarafPropertyCsvFactory or something else

File: extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+            }
+          }
+        } catch ( Exception e ) {
+          throw new IllegalStateException( getErrorPrefix() );
+        }
+      }
+
+    } catch ( FileNotFoundException e ) {
+      e.printStackTrace();
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java]: printStackTrace needs to be replaced with logging. Shouldn't we throw Exceptions here?

File: extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    } catch ( IOException e ) {
+      e.printStackTrace();
+    } finally {
+      if ( br != null ) {
+        try {
+          br.close();
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java]: You can use IOUtils.closeQuietly()

File: extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+import org.apache.commons.lang.StringUtils;
+
+public class KarafPropertyImporter {
+  final static String delimiter = ",";
+  protected static KarafInstance karafInstance;
+  private String importFilePath;
+  int lineNumber;
+
+  public KarafPropertyImporter( String importFilePath ) {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafPropertyImporter.java]: I would rather see this rewritten to use RegEx and capture groups within that to extract the values. At least with RegEx I can see the patterns and easily understand how the file should look. All of the split/indexOf/loop code is hard to follow

File: extensions/src/org/pentaho/platform/web/http/api/resources/KarafResource.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.pentaho.platform.osgi.KarafInstance;
+import org.pentaho.platform.osgi.KarafInstancePropertyDto;
+
+/**
+ * This service provides methods related to karaf instances
+ *
+ * @author tkafalas
+ */
+@Path( "/karaf" )
+public class KarafResource extends AbstractJaxRSResource {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/KarafResource.java]: Yea want to see this reimagined as a generic service for the platform. Anyone should be able to add a service/port registration to be returned by this service. Karaf would be just one of the systems feeding ports into it. So probably need a ServiceRegistry class and a ServiceResource. Add the ability to add a Service, with name, ID, Description and used ports

**Comment by pentaho-nbaker** [comment]
In general there needs to be more Javadoc

File: extensions/src/org/pentaho/platform/osgi/KarafInstance.java [hunk 8]
```diff
@@ -25,29 +48,60 @@
...
   private static final int MAX_NUMBER_OF_KARAF_INSTANCES = 50;
   private static final int BANNER_WIDTH = 79;
+  private final PortAssigner portAssigner = new PortAssigner();
+  private boolean started;
+  private static KarafInstance instance; // Not final because unit test creates multiple instances
+
   private StringBuilder banner = new StringBuilder();
 
-  public KarafInstance( String root ) {
+  protected KarafInstance( String root ) {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafInstance.java]: Just leave it package protected, not "protected" so that Unit Tests can call the constructor. Everyone else should be calling getInstance() (KarafBoot seems to be calling the constructor)

File: extensions/src/org/pentaho/platform/osgi/KarafInstance.java [hunk 9]
```diff
@@ -25,29 +48,60 @@
...
-    System.setProperty( "karaf.port", "" + assignedPorts[0] );// was 8081
-    System.setProperty( "org.pentaho.osgi.service.http.port", "" + assignedPorts[1] );// was 8181
-    System.setProperty( "org.pentaho.jmx.rmi.server.port", "" + assignedPorts[2] );// was 44444
-    System.setProperty( "org.pentaho.jmx.rmi.registry.port", "" + assignedPorts[3] );// was 1099
-    System.setProperty( "karaf.data", cachePath );
+    registerProperty( new KarafInstanceProperty( "CACHE_DIR", "karaf.data", "Karaf Cache Directory", cachePath ) );
+  }
+
+  public static KarafInstance getInstance() {
+    if ( instance == null ) {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/KarafInstance.java]: Create instance if null here

File: extensions/src/org/pentaho/platform/settings/ServerPortService.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * This serves to group {@link ServerPort} objects under an arbitrary name. It also allows a user friendly description
+ * of the service category to be supplied. Some web services will provide information on ServerPorts grouped by the
+ * {@link ServerPortService} they are under. To fully activate the ServerPortService it must be registered with
+ * {@link ServerPortRegistry#addService(ServerPortService)}
+ * 
+ * @author tkafalas
+ *
+ */
+public class ServerPortService {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/settings/ServerPortService.java]: This name is really confusing. I would think it's a service returning a ServicePort. Can we just make it Service?

File: extensions/src/org/pentaho/platform/web/http/api/resources/PortResource.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+import org.pentaho.platform.settings.ServerPortService;
+import org.pentaho.platform.settings.ServerPortServiceDto;
+
+/**
+ * This service provides methods related to karaf instances
+ *
+ * @author tkafalas
+ */
+@Path( "/port" )
+public class PortResource extends AbstractJaxRSResource {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/PortResource.java]: Should be a ServiceResource. We're returning a list of services, embedded within that are the ports being used.
